joomla/joomla-cms
# [4.0] Decoupling highlighter plugin from com_finder

This PR tries to fix #20252. 

This PR introduces a new event, "onFinderResult", which takes a single result object and the search query object. This allows plugins to modify the results (for example changing MIME types, link types, adding additional data, etc.) and in this case it allows the highlight plugin to modify the URL of the link of the search result to add the highlight parameter. This decouples the highlight system plugin from com_finder.

~~~The downside in this case is, that the URL, that is displayed below the search result, now also contains the highlight parameter. I would say that we should rework the output of the results anyway, but that would be the topic of another PR.~~~
The ```FinderIndexerResult``` object has both an ```url``` and ```route``` attribute and we are now using ```url``` for display purposes and ```route``` for the links href.
I also tried to change a few of the calls to the new namespaced versions.

[UPDATE]
In further reviews, I also found out the issue why searchwords with utf8 characters are not properly highlighted. It turns out that OutputFilter::stringJSSafe() is not utf8 aware and I had to rewrite it. Now it properly encodes this into a utf8 escaped string.

[UPDATE]
Yet another update regarding the ```url``` and ```route```: url is actually the ID of the item and thus can not be used. Instead I added ```cleanurl``` instead.

## How to test
* Test smart search and see the highlighting parameter in the URL of the search results.
* Apply the patch and see that it works as before, except that the additional highlight parameter is in the URL below the result.
* This PR mainly needs a codereview.
